### PR TITLE
Visualizer: add buttons for adding semantic action & semantic editor

### PR DESCRIPTION
**Pull request by lilyx**
@pdubroy : First step for adding semantic actions to visualizer, only have the `add` and `action name` buttons, no semantic editor yet. Just want to let you review it before your vacation.

### CODE REVIEW

File: visualizer/index.html [hunk 1]
```diff
@@ -165,6 +166,50 @@
...
+      resize: none;
+      outline: none;
+      padding-top: 1px;
+      padding-bottom: 1px;
+      margin-bottom: 4px;
+      height: 16px;
+      border: 1px solid #555;
+      border-radius: 10px;
+      display:block;
+      clear:both;
```
Comment (pdubroy) [inline visualizer/index.html]: I don't think you need this. `clear` is usually only used with floated elements.

File: visualizer/index.js [hunk 2]
```diff
@@ -140,6 +140,16 @@ function parseGrammar(source) {
     triggerRefresh(250);
   });
 
+  var actionContainers = document.querySelectorAll('.actionEntries');
+  Array.prototype.forEach.call(actionContainers, function(actionContainer) {
+    actionContainer.addEventListener('click', function() { triggerRefresh(); });
+    actionContainer.addEventListener('keypress', function(event) {
+      if (event.keyCode ===  13) {
```
Comment (pdubroy) [inline visualizer/index.js]: Nit: looks like you've got an extra space in there before the `13`.

File: visualizer/index.html [hunk 3]
```diff
@@ -46,6 +46,7 @@
     }
     #visualizerBody {
       flex-direction: column;
+      overflow: auto;
```
Comment (pdubroy) [inline visualizer/index.html]: I don't think you'll need this anymore. As I was looking at this pull request (and trying to see why you needed to add the `overflow` here), I realized that the addition of the `#visualizerBody` in your earlier PR caused some problems with the scrolling. I've just pushed a fix, so I think you can remove this line from your patch now.

File: visualizer/index.html [hunk 4]
```diff
@@ -165,6 +166,50 @@
       color: white;
     }
 
+    #actions {
+      display: flex;
+      flex: 1;
+      position: fixed;
```
Comment (pdubroy) [inline visualizer/index.html]: I think this should be `position: absolute` now.

File: visualizer/index.html [hunk 5]
```diff
@@ -165,6 +166,50 @@
...
+      right: 20px;
+      margin-top: 40px;
+    }
+    #actions .actionColumn {
+      margin-left: 15px;
+    }
+    #actions .actionEntries {
+      flex-direction: column;
```
Comment (pdubroy) [inline visualizer/index.html]: You don't need this. `flex-direction` is a property for _flex containers_ (i.e., elements that have `display: flex`), not _flex items_. Here's a good guide to flexbox if you haven't seen it: https://css-tricks.com/snippets/css/a-guide-to-flexbox/

File: visualizer/index.html [hunk 6]
```diff
@@ -165,6 +166,50 @@
...
+      align-self: flex-start;
```
Comment (pdubroy) [inline visualizer/index.html]: I think you also might not need this. At least, when I disabled it, it didn't seem to change anything.

File: visualizer/index.js [hunk 7]
```diff
@@ -140,6 +140,16 @@ function parseGrammar(source) {
     triggerRefresh(250);
   });
 
+  var actionContainers = document.querySelectorAll('.actionEntries');
```
Comment (pdubroy) [inline visualizer/index.js]: Could you put the following comment above this?

```
// TODO: Move this code into parseTree.js.
```

This code is really specific to the parseTree, but I see that you need it here to be able to call `triggerRefresh`. I think @mroeder's refactoring is going to make stuff like this easier.

**Comment by pdubroy** [comment]
LGTM with the comments I suggested.

File: visualizer/index.html [hunk 8]
```diff
@@ -161,6 +171,105 @@
...
+    .semanticsEditor button.save {
+      width: 100%;
+    }
+    .semanticsEditor button.save:hover {
+      text-decoration: underline;
+    }
+
+    #actions {
+      display: flex;
+      position: fixed;
```
Comment (lilyx) [inline visualizer/index.html]: Trying to keep the action buttons position fixed, so it won't change while we scrolling the bottom section

File: visualizer/index.html [hunk 9]
```diff
@@ -161,6 +171,105 @@
...
+      background-color: rgba(0, 0, 0, 0);
+    }
+    textarea.action {
+      resize: none;
+      outline: none;
+      margin: 4px 4px 4px 0;
+      height: 16px;
+      border: 1px solid #999;
+      border-radius: 10px;
+      display:block;
```
Comment (lilyx) [inline visualizer/index.html]: Missing a space, added in following commit

File: visualizer/index.html [hunk 10]
```diff
@@ -203,7 +312,7 @@
     #bottomSection {
       display: flex;
       flex: 1;
-      flex-direction: column;
```
Comment (lilyx) [inline visualizer/index.html]: Trying to list zoomOutButton, parse result, and action buttons side by side, so there won't be too much vertical spaces taken by bottoms.

File: visualizer/index.html [hunk 11]
```diff
@@ -203,7 +312,7 @@
...
+      overflow-x: auto;
```
Comment (lilyx) [inline visualizer/index.html]: Scroll horizontally for long expanded example.

File: src/Semantics.js [hunk 12]
```diff
@@ -380,6 +388,17 @@ Semantics.prototype.extendOperationOrAttribute = function(type, name, actionDict
   this[typePlural][name].checkActionDict(this.grammar);
 };
 
+Semantics.prototype.getOperationOrAttribute = function(operationOrAttributeName) {
+  if (operationOrAttributeName in this.attributes) {
+    return this.attributes[operationOrAttributeName];
+  } else if (operationOrAttributeName in this.operations) {
+    return this.operations[operationOrAttributeName];
+  } else {
+    throw new Error('Cannot find operation or attribute has name ' +
```
Comment (pdubroy) [inline src/Semantics.js]: I would just return undefined here, and let the calling code handle the error.

File: visualizer/index.html [hunk 13]
```diff
@@ -161,6 +171,105 @@
       color: white;
     }
 
+    .semanticsEditor.hidden .header,
+    .semanticsEditor.hidden .body,
+    .semanticsEditor.hidden .save {
+      display: none;
```
Comment (pdubroy) [inline visualizer/index.html]: Can you just give 'display: none' to the entire .semanticsEditor?

File: visualizer/index.html [hunk 14]
```diff
@@ -161,6 +171,105 @@
...
+    #actions .actionEntries {
+      display: flex;
+      flex-direction: column;
+      background-color: white;
+      color: black;
+      font-family: Menlo, Monaco, sans-serif;
+      font-style: normal;
+      border-radius: 3px;
+      padding: 2px 0px 2px 0px;
+      background-color: rgba(0, 0, 0, 0);
```
Comment (pdubroy) [inline visualizer/index.html]: Duplicate entry of `background-color`.

File: visualizer/index.html (hunk 11 above)
Comment (pdubroy) [inline visualizer/index.html]: There's something not quite right here. With a big example, I now see two horizontal scroll bars in the bottom pane.

File: visualizer/index.html (hunk 13 above)
Comment (lilyx) [inline visualizer/index.html]: For the editor that has classname 'hidden' I want to show the result only. I could see that 'hidden' is a confusing classname here, maybe change to 'resultOnly'?

File: visualizer/index.html (hunk 11 above)
Comment (lilyx) [inline visualizer/index.html]: Couldn't reproduce the problem. Could you specify which example you are using?

**Pull request closed without merging**